### PR TITLE
support override host status restriction for stateful session

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: support override host status restriction for stateful session
Additional Description:

Stateful session will try to parse upstream address from downstream request directly and override the result of load balancing algorithm by the `LoadBalancerContext::overrideHostToSelect` API.

To avoid the load balancer selecting hosts that in unexpected statuses, specifying some expected statuses are necessary.

In the previous design, we will provide expected statuses of override host by the `LoadBalancerContext::overrideHostToSelect` API.

And in the PR #18207, after some discussion with @htuch, we found may be cluster-level config may be more reasonable design and implementation.
Ref some more details: https://github.com/envoyproxy/envoy/pull/18207#discussion_r770235105

So this PR try to close previous discussion in the #18207:
- Refactoring `LoadBalancerContext::overrideHostToSelect` API to remove expected statuses for the return value.
- Add new common lb config `override_host_status` and related implementation.

Risk Level: Mid.
Testing: N/A.
Docs Changes: N/A.
Release Notes: N/A.
Platform Specific Features: N/A.

### CODE REVIEW

**Comment by wbpcode** [comment]
cc @htuch Could you please review this PR's new API when you have free time? 😄

**Comment by wbpcode** [comment]
cc @howardjohn

File: api/envoy/config/core/v3/health_check.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/health_check.proto:57]: Do you expect this message to have more fields?
If not, what's the advantage over just specifying the repeated field directly (in your specific case renaming the field to `override_host_status_set`)?

File: api/envoy/config/cluster/v3/cluster.proto
Comment (adisuissa) [inline api/envoy/config/cluster/v3/cluster.proto:606]: Is this field only relevant when the stateful_session extension is used?
If so, can you add this to the comment, and also add a reference to it?

File: api/envoy/config/core/v3/health_check.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/health_check.proto:59]: Could this be an empty set? If not, then add a PGV constraint.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Looked mostly at the API, left a few comments.

File: api/envoy/config/core/v3/health_check.proto
Comment (wbpcode) [reply api/envoy/config/core/v3/health_check.proto:57]: This is because we need to distinguish between the case where no override_host_status is configured at all and the case where 0 items are configured. The former will provide a default value.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (snowp) [reply api/envoy/config/cluster/v3/cluster.proto:606]: Should we be referring to the specific extension here at all? Or should we instead have this talk about the host override feature in general? I can imagine other extensions wanting to use host overrides as well

Comment (wbpcode) [reply api/envoy/config/cluster/v3/cluster.proto:606]: For the moment, it does. I will add some more comments. Thanks.

File: source/common/upstream/load_balancer_impl.cc
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:527]: Then `override_host_status` will be zero and all the override host from `LoadBalancerContext::overrideHostToSelect()` will be ignored and stateful session stickiness will be disabled for current cluster.

File: api/envoy/config/core/v3/health_check.proto
Comment (wbpcode) [reply api/envoy/config/core/v3/health_check.proto:59]: It can be an empty set which will disable stateful session stickiness for current cluster.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (adisuissa) [reply api/envoy/config/cluster/v3/cluster.proto:606]: One thing that I do not fully grok is what is expected if a config has this field set, but doesn't have the sticky_session extension configured.

Comment (wbpcode) [reply api/envoy/config/cluster/v3/cluster.proto:606]: > Should we be referring to the specific extension here at all? Or should we instead have this talk about the host override feature in general? I can imagine other extensions wanting to use host overrides as well

@htuch This configuration is designed for general host override. But `LoadBalancerContext::overrideHostToSelect()` is internaly C++ API which is annomymous for public users and the stateful session is the only related implemenatation for now.
May be we can ref the specific extension just as an example like:

```
The result of load balancing is allowed to be overwritten by the L4/L7 filters specified host.
For example the `stateful session filter <........>` will specfic upstream host directly accoriding to the 
downstream request.
But the overridden will only worked when the status of specficed host is as expected in the configuration.

```

Comment (htuch) [reply api/envoy/config/cluster/v3/cluster.proto:606]: I'm a bit lost reading this. I think there is context assumed of the reader that isn't likely to be there in most cases. Can you explain this section as if the reader had never heard of this override feature, pointing them to the relevant docs with RST links and/or some examples?

Comment (wbpcode) [reply api/envoy/config/cluster/v3/cluster.proto:606]: The override host feature is currently only available through the L4/L7 extensions. Now stateful session is the only extension that is ready.

So my intention is to use the stateful session as an example to help the reader understand the functionality and the configuration.

I will think about how to make it clearer and easier to understand. 🤔

File: api/envoy/config/cluster/v3/cluster.proto
Comment (adisuissa) [inline api/envoy/config/cluster/v3/cluster.proto:602]:
```suggestion
    // If this is set and and has no item then the override host will be ignored.
```

**Review by adisuissa — COMMENTED** [review]
Minor comment about the api, but otherwise LGTM.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (wbpcode) [reply api/envoy/config/cluster/v3/cluster.proto:602]: get it.

**Comment by wbpcode** [comment]
@adisuissa @htuch docs/API is updated. Thanks for all your patient reivews. And sorry for the delayed replies and updates.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by wbpcode** [comment]
hi, @snowp , the API is ready and could you give a review to the implementation when you have free time? Thanks. 🌷

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by KBaichoo** [comment]
ping @snowp  PTAL :)

**Review by snowp — COMMENTED** [review]
Few questions to hopefully make the docs a bit clearer, impl wise this looks good

**Comment by snowp** [comment]
FYI I will be out until March 3rd, so I'm likely not going to get to review again until then. Implementation wise this seems fine, just want to make sure the documentation ends up being easy to understand. 

@htuch @adisuissa Do you think either of you can handle merging this once comments are addressed?

**Comment by adisuissa** [comment]
> FYI I will be out until March 3rd, so I'm likely not going to get to review again until then. Implementation wise this seems fine, just want to make sure the documentation ends up being easy to understand.
> 
> @htuch @adisuissa Do you think either of you can handle merging this once comments are addressed?

Sure @snowp I can take this one once your comments are addressed.
Thanks, and enjoy your time off.

**Comment by htuch** [comment]
I can provide senior maintainer coverage once @adisuissa has reviewed.

File: api/envoy/config/core/v3/health_check.proto
Comment (wbpcode) [reply api/envoy/config/core/v3/health_check.proto:59]: @snowp Yep. If override_host_status is set with an empty list of statuses then the override host will be ignored. And if override_host_status is not set completely, we will give some default statuses.

File: docs/root/intro/arch_overview/upstream/load_balancing/override_host.rst
Comment (wbpcode) [reply docs/root/intro/arch_overview/upstream/load_balancing/override_host.rst:8]: get it.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (wbpcode) [reply api/envoy/config/cluster/v3/cluster.proto:604]: get it.

**Comment by wbpcode** [comment]
Hi, @adisuissa the review comments from @snowp are addressed. PTAL, thanks. 🌷

File: source/common/upstream/load_balancer_impl.cc
Comment (adisuissa) [inline source/common/upstream/load_balancer_impl.cc:548]: I think this may hide a case where the server sends a config with some other health status.
Consider adding `[(validate.rules).enum.defined_only = true]` to the `statuses`, and maybe a emit a log here.

**Review by adisuissa — CHANGES_REQUESTED** [review]
Thanks! Overall LGTM.
Left one minor comment.
/wait

File: source/common/upstream/load_balancer_impl.cc
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:548]: get it.

**Comment by rojkov** [comment]
@adisuissa gentle ping

**Review by adisuissa — DISMISSED** [review]
/lgtm api
LGTM code changes.
@htuch for final review.
Also, not sure if this needs release notes.

**Review by htuch — CHANGES_REQUESTED** [review]
Looks good, but a few suggestions.

File: source/common/upstream/load_balancer_impl.cc
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:548]: get it.

File: source/common/upstream/load_balancer_impl.cc
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:28]: @htuch 🤔 We don't have complex bit operations here, so there doesn't seem to be any particular benefit to using `std::bitset`.

**Review by adisuissa — APPROVED** [review]

File: source/common/upstream/load_balancer_impl.cc
Comment (htuch) [reply source/common/upstream/load_balancer_impl.cc:28]: It's not about complexity of bit operations, more about doing this in a more native C++ way; ideally use the right ADTs rather than bit twiddling, which should be less error prone.

Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:28]: get it.

File: source/common/upstream/load_balancer_impl.cc
Comment (htuch) [inline source/common/upstream/load_balancer_impl.cc:518]: Nit: can you add a `using` type alias for this, e.g. `using HostStatusSet = std::bitset<32>`? search-replace this across the diff.

File: source/common/upstream/load_balancer_impl.h
Comment (htuch) [inline source/common/upstream/load_balancer_impl.h:168]: FWIW it doesn't need to be 32-bit long, but it probably doesn't matter that much either.

**Review by htuch — COMMENTED** [review]
LGTM modulo one last nit on type alias.

File: source/common/upstream/load_balancer_impl.h
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.h:168]: Yes. Only 3 bit is used for now. But in the x64 matchine, the storage size of std::bitset<N> (N > 0) is at least 8 bytes. So I guess it's not a big problem. So I think setting N to 32 or 8 makes little difference. 

```
    std::cout << sizeof(std::bitset<8>) << std::endl;  // 8
    std::cout << sizeof(std::bitset<16>) << std::endl;  // 8
    std::cout << sizeof(std::bitset<32>) << std::endl;  // 8
    std::cout << sizeof(std::bitset<64>) << std::endl;  // 8
    std::cout << sizeof(std::bitset<128>) << std::endl; // 16
```

File: source/common/upstream/load_balancer_impl.cc
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:518]: get it.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
